worlds/cv64/__init__.py

```diff
@@ -0,0 +1,330 @@
+            offset_data.update(get_loading_zone_bytes(self.options, self.starting_stage, self.active_stage_exits))
+            # Countdown
+            if self.options.countdown.value:
+                offset_data.update(get_countdown_numbers(self.options, active_locations))
+            # Start Inventory
+            offset_data.update(get_start_inventory_data(self.options, self.options.start_inventory.value, self.options.start_inventory_from_pool.value))
+
+            cv64_rom = LocalRom(get_base_rom_path())
+
+            slot_name = self.multiworld.player_name[self.player].encode("utf-8")
+
+            rompath = os.path.join(output_directory, f"{self.multiworld.get_out_file_name_base(self.player)}.z64")
+
+            patch_rom(self.multiworld, self.options, cv64_rom, self.player, offset_data, self.active_stage_exits,
+                      self.s1s_per_warp, self.active_warp_list, self.required_s2s, self.total_s2s, shop_name_list,
+                      shop_desc_list, shop_colors_list, slot_name, active_locations)
+
+            cv64_rom.write_to_file(rompath)
+
+            patch = CV64DeltaPatch(os.path.splitext(rompath)[0] + CV64DeltaPatch.patch_file_ending, player=self.player,
+                                   player_name=self.multiworld.player_name[self.player], patched_path=rompath)
+            patch.write()
+        except:
+            print(f"D'oh, something went wrong in CV64's generate_output for player "
+                  f"{self.multiworld.player_name[self.player]}!")
+            raise
+        finally:
+            if os.path.exists(rompath):
+                os.unlink(rompath)
+            self.rom_name_available_event.set()  # make sure threading continues and errors are collected
+
+    def write_spoiler(self, spoiler_handle: typing.TextIO) -> None:
+        # Write the stage order to the spoiler log
+        spoiler_handle.write(f"\nCastlevania 64 stage & warp orders for {self.multiworld.player_name[self.player]}:\n")
+        for stage in self.active_stage_list:
+            num = str(self.active_stage_exits[stage]["position"]).zfill(2)
+            path = self.active_stage_exits[stage]["path"]
+            spoiler_handle.writelines(f"Stage {num}{path}:\t{stage}\n")
+
+        # Write the warp order to the spoiler log
+        spoiler_handle.writelines(f"\nStart :\t{self.active_stage_list[0]}\n")
+        for i in range(1, len(self.active_warp_list)):
+            spoiler_handle.writelines(f"Warp {i}:\t{self.active_warp_list[i]}\n")
+
+    def fill_slot_data(self) -> typing.Dict[str, typing.Any]:
+        return {"death_link": self.options.death_link.value}
+
+    def modify_multidata(self, multidata: dict):
+        # wait for self.rom_name to be available.
+        self.rom_name_available_event.wait()
+        rom_name = getattr(self, "rom_name", None)
+        # we skip in case of error, so that the original error in the output thread is the one that gets raised
+        if rom_name:
```

Comment:
I decided to remove the modify_multidata stuff entirely, since out of all the things I copied from DKC3 it's probably the one I understand the least. Nothing seems to blow up without it as-is.

What other things would you recommend I use beside the slot name?

Reply:
The ROM verification should now be more like how it is in LADX and Emerald. Instead of the player's name, it's now a 16-byte number that is generated in generate_early that I use as the authentication key. Also, I added that extra step from Emerald for checking the patch version in addition to whether the ROM is patched or not.